### PR TITLE
cody-gateway: wrap events logging in buffer

### PR DESCRIPTION
**Pull request by bobheadxi**
This change moves events logging out of the request hotpath so that we are not blocking request handling on event logging. This should improve per-request latency by ~30ms to ~100ms, based on an unscientific random sampling of request traces over the last few weeks (Cloud Trace doesn't make it easy to perform analysis over specific span names).

I also made a minor refactoring of instrumentation by moving it to its own layer for ease of reuse and consistency.

Closes https://github.com/sourcegraph/sourcegraph/issues/52033

## Test plan

Unit tests. I also added a `CODY_GATEWAY_BUFFERED_LAGGY_EVENT_LOGGING_FUN_TIMES_MODE` (intentionally not registered in config) to demonstrate this IRL - in the trace below, the request concludes long before our laggy event logging concludes:

image

### CODE REVIEW

File: enterprise/cmd/cody-gateway/internal/events/buffered.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// in the background instead of in the hot-path of a request. It implements
+// goroutine.BackgroundRoutine that must be started.
+func NewBufferedLogger(logger log.Logger, handler Logger, queueSize int) *BufferedLogger {
+	return &BufferedLogger{
+		log: logger.Scoped("bufferedLogger", "buffered events logger"),
+
+		handler: handler,
+
+		bufferC:      make(chan bufferedEvent, queueSize),
+		timeout:      100 * time.Millisecond,
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/events/buffered.go:49]: super nit: maybe a constant at the top of the file would be helpful to easily find where to tweak this

File: enterprise/cmd/cody-gateway/internal/events/buffered.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+	case <-time.After(l.timeout):
+		return errors.Newf("failed to insert event in %s: buffer full: %d items pending",
+			l.timeout.String(), len(l.bufferC))
+	}
+}
+
+// Start starts buffered logger's background processing job.
+func (l *BufferedLogger) Start() {
+	go func() {
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/events/buffered.go:75]: a goroutine.BackgroundRoutine doesn't need another go routine wrap, does it?

File: enterprise/cmd/cody-gateway/internal/events/buffered.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		for event := range l.bufferC {
+			if err := l.handler.LogEvent(event.spanCtx, event.Event); err != nil {
+				l.log.Error("failed to log buffered event", log.Error(err))
+			}
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/events/buffered.go:79]: would be slick if we could batch the insertions here by deferring up to Xms until we actually log them 😬 But that's harder, so feel free to file a ticket that we look at it eventually instead.

File: enterprise/cmd/cody-gateway/internal/events/buffered.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		l.log.Info("events buffer emptied")
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/events/buffered.go:82]: Maybe? That is a bit more informational to me 

```suggestion
		l.log.Info("all events flushed")
```

File: enterprise/cmd/cody-gateway/internal/events/buffered.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		return l.handler.LogEvent(spanCtx, event)
+	}
+
+	select {
+	case l.bufferC <- bufferedEvent{spanCtx: spanCtx, Event: event}:
+		return nil
+
+	case <-time.After(l.timeout):
+		return errors.Newf("failed to insert event in %s: buffer full: %d items pending",
+			l.timeout.String(), len(l.bufferC))
```
Comment (eseliger) [inline enterprise/cmd/cody-gateway/internal/events/buffered.go:69]: I'm a bit scared that we are failing to log events here 😞 we use those for usage charts at the moment and while they're not billing critical, it would also be sad to lose a bunch of events. I would either want us to ramp the timeout a bunch, or rely on that the draining on the other end is gonna be fast enough that it will eventually enqueue and not timeout here. WDYT?

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:75]: What's the right way to implement this? The interface only has a start and stop

Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:75]: I think start can be blocking, so just removing the go routine in here should be fine. The MonitorBackgroundRoutines should handle that well.

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 3 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:79]: I think that increases the risk of losing an event, and might be a bit of a premature optimization? The submissions are fast, just not fast enough to justify keeping in the request hot path

Comment (eseliger) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:79]: alright, I was thinking that backpressure might be resolved easier if we bulk inserted them, but no evidence that it would so feel free to ignore me

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 5 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:69]: It is _very_ unlikely to get blocked unless events were never getting submitted in the first place IMO, and we have some knobs (buffer size) that we can adjust live. I can increase both default buffer size and timeout for now though!

File: enterprise/cmd/cody-gateway/shared/config.go [hunk 6]
```diff
@@ -79,6 +81,7 @@ func (c *Config) Load() {
 	c.BigQuery.ProjectID = c.Get("CODY_GATEWAY_BIGQUERY_PROJECT_ID", os.Getenv("GOOGLE_CLOUD_PROJECT"), "The project ID for the BigQuery events.")
 	c.BigQuery.Dataset = c.Get("CODY_GATEWAY_BIGQUERY_DATASET", "CODY_GATEWAY", "The dataset for the BigQuery events.")
 	c.BigQuery.Table = c.Get("CODY_GATEWAY_BIGQUERY_TABLE", "events", "The table for the BigQuery events.")
+	c.BigQuery.EventBufferSize = c.GetInt("CODY_GATEWAY_BIGQUERY_EVENT_BUFFER_SIZE", "10", "The number of events to buffer when submitting BigQuery events - set to 0 to disable.")
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/shared/config.go]: Increased!

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:79]: Ah right! 😮‍💨

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 5 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Additionally, the error returned is logged as errors at callsites so we'll know very quickly if there's an issue

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 3 above)
Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go]: Let's defer this for now, we can observe the lag time live via our traces

**Review by eseliger — APPROVED** [review]

File: enterprise/cmd/cody-gateway/internal/events/buffered.go (hunk 5 above)
Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Second the concern here... we use the buffered event logging as an optimization for the happy path, but itself isn't a means to event logging.

Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Sorry, can you clarify? I'm not sure I follow

Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Sorry, let me try again :D 

First to clarify my understanding, if the queue gets stuck after timeout (e.g. 150ms), then events get dropped, is this correct?

If so, then we would lose those events, instead of eventually getting them logged (i.e. send to BigQuery). 

By "an optimization for the happy path, but itself isn't a means to event logging", I mean we do not want the action of sending events to BigQuery to become a performance bottleneck in normal circumstances, which is our goal IIUC.

However if we start dropping events because it's not sending fast enough, that makes the buffered logger become the "means", not just "an optimization" anymore (not our goal).

Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Kind of - we can switch this to log in the hotpath if the buffer fills up, similar to the approach I ended up adding here where if something comes in after shutdown we submit it immediately

Comment (unknwon) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: Yeah, that would be great, because after all we do want these events, right 😁

Comment (bobheadxi) [reply enterprise/cmd/cody-gateway/internal/events/buffered.go:77]: https://github.com/sourcegraph/sourcegraph/pull/53282
